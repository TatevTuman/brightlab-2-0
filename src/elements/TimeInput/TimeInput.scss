@import "static/assets/styles/mixins";

.time {
  display: flex;
  position: relative;
  max-height: toRem(48);
  max-width: toRem(160);
  padding: toRem(10) 0;
  border: 1px solid var(--white-p);
  background: var(--white-p);
  mix-blend-mode: normal;
  border-radius: toRem(10);
  cursor: pointer;

  &-label {
    position: absolute;
    bottom: toRem(55);
    color: var(--primary);
    font-weight: 700;
    text-align: center;

    & > div {
    }
  }

  &:active, &:focus-within {
    border: 1px solid rgba(62, 66, 223, 0.24);
    background: var(--white);
    outline: none;
  }

  div {
    width: 50%;
    position: relative;
    text-align: center;
    padding-left: toRem(15);
    background: transparent;
    border: none;
    mix-blend-mode: normal;
    font-size: toRem(15);
    font-weight: 500;
    color: #1D2069;
    opacity: 0.7;

    input {
      flex: 1;
      border: none;
      height: 100%;
      width: toRem(60);
      overflow: scroll;
      font-size: toRem(15);
      color: var(--black);
      background: transparent;
    }
  }
  div:first-child {
    border-right: 1px solid rgba(62, 66, 223, 0.24);
  }
  div:last-child {
    padding-right: toRem(15);
  }
}