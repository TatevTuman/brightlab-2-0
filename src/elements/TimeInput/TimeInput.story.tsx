import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import TimeInput, { TimeInputProps } from './TimeInput'
import { action } from '@storybook/addon-actions'
import { storybook } from '@utils'

export default {
  title: 'Elements/TimeInput',
  component: TimeInput,
  argTypes: {
    label: { control: 'text', name: 'Label' },
    ...storybook.controls.interactive
  }
} as Meta

const TimeInputStoryTemplate: Story<TimeInputProps> = args => {
  return <TimeInput {...args} />
}

TimeInputStoryTemplate.args = {
  ...storybook.actions.interactive
}

export const TimeInputStory = TimeInputStoryTemplate.bind({})
TimeInputStory.args = {
  ...TimeInputStoryTemplate.args
}
