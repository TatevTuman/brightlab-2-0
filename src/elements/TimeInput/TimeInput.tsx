import React, { memo, useState, useEffect, useRef } from 'react'
import './TimeInput.scss'

export interface TimeInputProps {
  onChange: (value: string) => void
  label?: string
}

enum ArrowKeys {
  Right = 'ArrowRight',
  Left = 'ArrowLeft'
}

const TimeInput: React.FC<TimeInputProps> = props => {
  const { label, onChange } = props
  const [time, setTime] = useState({
    firstInputValue: '',
    secondInputValue: ''
  })
  // console.log(time)
  const firstRef: any = useRef<HTMLInputElement>(null)
  const secondRef: any = useRef<HTMLInputElement>(null)

  useEffect(() => {
    onChange && onChange(time.firstInputValue)
  }, [time.firstInputValue])

  useEffect(() => {
    onChange && onChange(time.secondInputValue)
  }, [time.secondInputValue])

  const handleFirstInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const firstCurrentRef: any = firstRef.current
    const firstCurrentRefValue = firstCurrentRef.value
    const maxLength = firstCurrentRef.maxLength
    if (e.target.value < '24') {
      setTime({ firstInputValue: e.target.value })
      if (firstCurrentRefValue.length === maxLength) {
        ;(secondRef.current as HTMLInputElement).focus()
      }
    }
  }

  const handleSecondInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const secondCurrentRef: any = secondRef.current
    const secondCurrentRefValue = secondCurrentRef.value
    if (e.target.value < '60') {
      setTime({ secondInputValue: e.target.value })
    }
    if (secondCurrentRefValue.length === 0) {
      ;(firstRef.current as HTMLInputElement).focus()
    }
  }

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === ArrowKeys.Right) {
      if (firstRef.current.value.length == 2) {
        ;(secondRef.current as HTMLInputElement)?.focus()
      }
    }

    if (e.key === ArrowKeys.Left) {
      if (secondRef.current.value.length == 0) {
        ;(firstRef.current as HTMLInputElement)?.focus()
      }
    }
  }

  return (
    <>
      <label className={'time'}>
        {label && (
          <label htmlFor={'HH'} className={'time-label'}>
            {label}
          </label>
        )}
        <div>
          <input
            name={'HH'}
            type={'text'}
            placeholder={'HH'}
            maxLength={2}
            value={time.firstInputValue}
            ref={firstRef}
            onChange={e => handleFirstInputChange(e)}
            onKeyDown={handleKeyDown}
          />
        </div>
        <div>
          <input
            name={'mm'}
            type={'text'}
            placeholder={'mm'}
            maxLength={2}
            value={time.secondInputValue}
            ref={secondRef}
            onChange={e => handleSecondInputChange(e)}
            onKeyDown={handleKeyDown}
          />
        </div>
      </label>
    </>
  )
}

export default memo(TimeInput)
