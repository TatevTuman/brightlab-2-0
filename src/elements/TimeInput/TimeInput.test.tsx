import React, { useState } from 'react'
import { render, cleanup, waitFor } from '@testing-library/react'
import { act, renderHook } from '@testing-library/react-hooks'
import TimeInput from './TimeInput'
import userEvent from '@testing-library/user-event'

afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('TimeInput', () => {
  const props = {
    label: 'test',
    onChange: jest.fn()
  }
  it('renders correctly with props', async () => {
    const { container } = render(<TimeInput {...props} />)
    expect(container).toMatchSnapshot()
  })
  it('renders correctly label', async () => {
    const { getByText } = render(<TimeInput {...props} />)
    expect(getByText(props.label)).toBeInTheDocument()
  })

  it('handles onChange', async () => {
    const { container, rerender } = render(<TimeInput {...props} />)
    const inputs = container.querySelectorAll('input')
    inputs.forEach((input, index) => {
      act(() => {
        userEvent.type(input, index.toString())
      })
      rerender(<TimeInput {...props} />)
    })
    expect(props.onChange).toHaveBeenCalledTimes(5)
  })

  it('focuses by tab', async () => {
    const { container } = render(<TimeInput {...props} />)
    await waitFor(() => container)

    const inputs = container.querySelectorAll('input')
    inputs.forEach((input, index) => {
      act(() => {
        userEvent.tab(input, index.toString())
        expect(input).toHaveFocus()
      })
    })
  })
  it('focuses by type', async () => {
    const { container } = render(<TimeInput {...props} />)
    await waitFor(() => container)
    const inputs = container.querySelectorAll('input')
    inputs.forEach((input, index) => {
      if (index === 0) {
        act(() => {
          userEvent.type(input, index.toString())
          expect(input).toHaveFocus()
        })
      } else if (index === 1) {
        act(() => {
          userEvent.type(input, index.toString())
          expect(input).toHaveFocus()
        })
      }
    })
  })
})
