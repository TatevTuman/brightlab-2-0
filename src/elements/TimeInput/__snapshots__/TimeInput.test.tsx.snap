// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeInput renders correctly with props 1`] = `
<div>
  <label
    class="time"
  >
    <label
      class="time-label"
      for="HH"
    >
      test
    </label>
    <div>
      <input
        maxlength="2"
        name="HH"
        placeholder="HH"
        type="text"
        value=""
      />
    </div>
    <div>
      <input
        maxlength="2"
        name="mm"
        placeholder="mm"
        type="text"
        value=""
      />
    </div>
  </label>
</div>
`;
